### PR TITLE
Unit tests for pimpl use case

### PR DESCRIPTION
**Pull request by Twon**
Fleshing out the pimpl usecase via tests

### CODE REVIEW

File: example_pimpl.cpp [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    assert(!pimpl_->get_name().empty());
+    return pimpl_->get_name().c_str();
+}
+
+void example_pimpl::set_name(char const * const name)
+{
+    assert(name);
+    pimpl_->set_name(name);
+}
+
```
Comment (Twon) [inline example_pimpl.cpp]: Note the below line force instantiation of the rule of zero functions in the translation unit where the pimpl inner class is defined.  This issue already exists for pimpl using std::unique_ptr, though in this case, you would only need to do so for the destructor.  Details on this in a Herb Sutter article here: https://herbsutter.com/gotw/_101/

File: indirect_value.h [hunk 2]
```diff
@@ -64,7 +64,7 @@ class indirect_value : private indirect_value_base<T, C> {
         ptr_ = std::unique_ptr<T, D>(get_c()(*i.ptr_), D{});
       }
       else{
-        *ptr_ = *i.ptr_;
+        ptr_.reset( get_c()(*i.ptr_) );
```
Comment (Twon) [inline indirect_value.h]: This is just to get this compiling temporarily, we know we have to update this method to take an assigner object already.

**Comment by Twon** [comment]
Hmmm... these tests do not fail locally for me.  The error suggests the version of libstdc++ that is being used it using reference counted strings.  Google suggests this was the case in version 4 of libstdc++ but I need to either debug on Travis to confirm or disable the failing statements if this is valid way the strings to behave.

File: indirect_value.h (hunk 2 above)
Comment (jbcoe) [reply indirect_value.h]: Smart move not getting blocked.

**Pull request merged**